### PR TITLE
feat(usescrollsync): add scroll syncing via hook: `useScrollSync`

### PR DESCRIPTION
**Pull request by pcattori**
@okonet : I took a shot at https://github.com/okonet/react-scroll-sync/issues/36#issuecomment-639553613 .

Right now I wrote this as a stand-alone, typed React hook. Ideally, if we agree on the hook, we could expose it as a feature via `import {useScrollSync} from 'react-scroll-sync'`, but additionally the existing `ScrollSync` and `ScrollSyncPane` component could optionally be refactored to make use of `useScrollSync` (perhaps via context API).

**Looking for feedback** to see if this is on the right path before continuing to polish this PR.

[CodeSandbox Example](https://codesandbox.io/s/usescrollsync-9j2ox?file=/src/useScrollSync.ts) : syncing scroll of the red header and the white body on horizontal axis.

Relates to #22, #36

---

Example usage:

```tsx
const MyComponent: React.FC = () => {
  const [thead, theadRef] = useElement()
  const [tbody, tbodyRef] = useElement()
  useScrollSync([thead, tbody], { axis: 'horizontal' })
  return (
    <>
      <div ref={theadRef} role="rowgroup" style={{ overflowX: 'auto' }}>{...}</div>
      <div ref={tbodyRef} role="rowgroup" style={{ overflow: 'auto' }}>{...}</div>
    </>
  )
}
```

Other changes:
- `axis` (typed as `'horizontal' | 'vertical' | 'both'`) to replace `horizontal=true` and `vertical=true` options. This allows user to only enable the axes they _want_, as opposed to having to disable the axes they don't want.
- Removed concept of "groups"; instead just use multiple calls of `useScrollSync`, one for each group.
- Included a `useElement` hook for convenience. It just copies the [recommended approach from the React docs](https://reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node) and wraps it in a hook.

### CODE REVIEW

**Comment by okonet** [comment]
I like this direction very much! Do you want to drop components in favor of hooks and do a breaking change? Or do you want to support both?

File: src/example-useScrollSync.tsx
Comment (okonet) [inline src/example-useScrollSync.tsx:6]: No need to type it. It’s discouraged by React team afaik.

File: src/useScrollSync.ts
Comment (okonet) [inline src/useScrollSync.ts:5]: Maybe also an enum? “proportional” | “absolute”?

File: src/useScrollSync.ts
Comment (okonet) [inline src/useScrollSync.ts:16]: One loop should be enough here. No need for this map. Just access the `.current` directly.

File: src/useScrollSync.ts
Comment (okonet) [inline src/useScrollSync.ts:17]: Should be a `forEach` since you not returning.

File: src/useScrollSync.ts
Comment (okonet) [inline src/useScrollSync.ts:30]: Missing dependencies here. When refs will change this effect won’t run.

File: src/useScrollSync.ts
Comment (okonet) [inline src/useScrollSync.ts:48]: Same as before: `forEach`

File: src/useScrollSync.ts
Comment (okonet) [inline src/useScrollSync.ts:89]: Why is linter has to be disabled?

**Review by okonet — CHANGES_REQUESTED** [review]
Good direction!

File: src/useScrollSync.ts
Comment (pcattori) [reply src/useScrollSync.ts:16]: I actually have a local implementation that makes the ref handling much simpler and should address this. Namely, the `useScrollSync` hook will just accept a list of `HTMLElement` (rather than a list of refs to `HTMLElement`s).

I'll update with this change soon.

File: src/useScrollSync.ts
Comment (pcattori) [reply src/useScrollSync.ts:89]: Not sure actually. I copied most of this from the [existing implementation](https://github.com/okonet/react-scroll-sync/blob/master/src/ScrollSync.js#L119). I tried removing the `eslint-disable-line`, but my editor keeps adding it back in... I'll investigate to see if it's just my editor config or something else.

**Comment by pcattori** [comment]
> I like this direction very much! Do you want to drop components in favor of hooks and do a breaking change? Or do you want to support both?

I think it makes sense to keep the components, at least temporarily. Mostly because there may be use-cases that rely on features from the component implementation (e.g. I don't implement the `onSync` callback yet).

Options:
1. Keep the components. Support both components and hooks.
2. Same as (1), but deprecate the components. This will give us some time to get feedback for bugs/features for the hook implementation.
3. Remove the components.

I think I lean towards (2). What do you think @okonet ?

File: src/example-useScrollSync.tsx
Comment (pcattori) [reply src/example-useScrollSync.tsx:6]: Good point! Especially since `App` doesn't make use of `children`, which `React.FC` implies.

So I agree! 😄 

Out of curiosity, do you have a link to the React team discouraging this? I'm interested in how they phrased it.

File: src/useScrollSync.ts
Comment (pcattori) [reply src/useScrollSync.ts:5]: Something like `mode: 'proportional' | 'absolute'`?

Does it make sense to default this to `'proportional'` or `'absolute'`? I know the current implementation defaults to `proportional=true`, but I feel like `'absolute'` might be a more natural default.

Actually, I might be most in favor of not defaulting this to anything at all and requiring user to specify.

File: src/useScrollSync.ts
Comment (pcattori) [reply src/useScrollSync.ts]: done

File: src/useScrollSync.ts
Comment (pcattori) [reply src/useScrollSync.ts]: done

File: src/useScrollSync.ts
Comment (pcattori) [reply src/useScrollSync.ts:44]: done

File: src/useScrollSync.ts
Comment (pcattori) [reply src/useScrollSync.ts]: 🤷 not sure why the linter was being disabled before. I removed the `eslint-disable-line` comment and things seem to do just fine. so this should now be addressed.

File: src/useScrollSync.ts
Comment (pcattori) [reply src/useScrollSync.ts]: I've updated the PR to include this new implementation. For now just adding as another commit, but if we get to the point where we want to merge it, I'm happy to squash/rebase commits as needed.

File: src/useScrollSync.ts
Comment (okonet) [reply src/useScrollSync.ts:4]: Yes, exactly. I think defaulting to `proportional` is more appropriate but I'm open for making it a required prop.

**Review by okonet — COMMENTED** [review]
Sorry I lost the track on this one. It looks good now so let's resolve the last comment. Also, the documentation needs to be updated.

File: src/example-useScrollSync.tsx
Comment (okonet) [reply src/example-useScrollSync.tsx]: I can't remember where I read about it TBH.

**Comment by okonet** [comment]
@pcattori I resolved almost all comments and awaiting for the final changes. Also, could you document the hook / update the documentation to reflect the new state?

I would say, let's remove components since maintenance efforts are way too high. Also documenting both approaches going to take longer. React isn't going back to pre-hook era anyways and we'll release this as a breaking change (we could even change the name of the project and package to reflect the change). WYT?

**Comment by pcattori** [comment]
> @pcattori I resolved almost all comments and awaiting for the final changes. Also, could you document the hook / update the documentation to reflect the new state?
> 
> I would say, let's remove components since maintenance efforts are way too high. Also documenting both approaches going to take longer. React isn't going back to pre-hook era anyways and we'll release this as a breaking change (we could even change the name of the project and package to reflect the change). WYT?

@okonet sounds good! I'll take a look at this over the weekend to finalize things.

**Comment by pcattori** [comment]
@okonet: I've implemented the `mode` parameter, added jsdoc for `useScrollSync`, and reworked the example into a markdown usage guide.

I think we should be good to go with that! Let me know if we need anything else before merging.

Next steps:
- Remove the existing components to decrease maintenance overhead

File: src/useScrollSync.ts
Comment (pcattori) [reply src/useScrollSync.ts]: Made `mode` a required parameter for now.

File: src/useScrollSync.ts
Comment (okonet) [inline src/useScrollSync.ts:4]: I think it’s odd to make `Mode` a required option but `axis` not. I would suggest to make both options optional (hence the name) and use same defaults as the components do. Alternatively, if you think it’s better, let’s make both options required. WYT?

**Review by okonet — CHANGES_REQUESTED** [review]
Let’s decide on both options to be options or required and then it’s good to go! Great work!

File: src/useScrollSync.ts
Comment (pcattori) [reply src/useScrollSync.ts]: Yes, I agree. I've made the relevant changes now. Let me know what you think.

**Comment by okonet** [comment]
Looking great! I'd say, let's release it as a minor version.

**Comment by okonet** [comment]
Oh one last thing, we should add it to the documentation website.

For that we'd need to modify the glob in the https://github.com/okonet/react-scroll-sync/blob/master/styleguide.config.js to include the hook as well. WYT?

You can preview it here https://deploy-preview-54--react-sync-scroll.netlify.app

**Comment by jayhuang-7** [comment]
Really appreciate your work, everyone.
but it seems like some type checking in TS are not passing. `useElement` is returning `CallbackRef<T | undefined>` but it is not compatible with `RegacyRef<T | null>`.

**Comment by pcattori** [comment]
> @okonet, @pcattori Really appreciate your work, everyone.
> but it seems like some type checking in TS are not passing. `useElement` is returning `CallbackRef<T | undefined>` but it is not compatible with `RegacyRef<T | null>`.

@jayhuang-7 : could you provide an example where this happens?

**Comment by jayhuang-7** [comment]
@pcattori 
I just followed the Readme.md like this one.

```javascript
const MyComponent: React.FC = () => {
  const [thead, theadRef] = useElement()
  const [tbody, tbodyRef] = useElement()
  useScrollSync([thead, tbody], { axis: 'horizontal' })
  return (
    <>
      <div ref={theadRef} role="rowgroup" style={{ overflowX: 'auto' }}>{...}</div>
      <div ref={tbodyRef} role="rowgroup" style={{ overflow: 'auto' }}>{...}</div>
    </>
  )
}
```
However, it seems that we can't set `theadRef` and `tbodyRef` in line 7 and 8 due to type error that I indicated the above.

**Pull request closed without merging**
